Source/Metal/CommandBufferMTL.mm

```diff
@@ -0,0 +1,481 @@
+    MTLRenderPassDescriptor* renderPassDesc = [MTLRenderPassDescriptor renderPassDescriptor];
+    
+    for(uint32_t i = 0; i < attachmentsDesc.colorNum; i++) {
+        DescriptorMTL& descriptorMTL = *(DescriptorMTL*)attachmentsDesc.colors[i];
+        
+        renderPassDesc.colorAttachments[i].texture = descriptorMTL.GetTextureHandle();
+        renderPassDesc.colorAttachments[i].clearColor = MTLClearColorMake(0, 0, 0, 1);
+        renderPassDesc.colorAttachments[i].loadAction = MTLLoadActionLoad;
+        renderPassDesc.colorAttachments[i].storeAction = MTLStoreActionStore;
+    }
+    
+    if(attachmentsDesc.depthStencil) {
+        DescriptorMTL& descriptorMTL = *(DescriptorMTL*)attachmentsDesc.depthStencil;
+        renderPassDesc.depthAttachment.texture = descriptorMTL.GetTextureHandle();
+        renderPassDesc.depthAttachment.loadAction = MTLLoadActionLoad;
+        renderPassDesc.depthAttachment.storeAction = MTLStoreActionStore;
+
+    }
+    //m_RenderPassDescriptor = [renderPassDescriptor copy];
+    m_RendererEncoder = [m_Handle renderCommandEncoderWithDescriptor: renderPassDesc];
+}
+
+void CommandBufferMTL::EndRendering() {
+    EndCurrentEncoders();
+}
+
+void CommandBufferMTL::SetViewports(const Viewport* viewports, uint32_t viewportNum) {
+    Scratch<MTLViewport> mtlViewports = AllocateScratch(m_Device, MTLViewport, viewportNum);
+    for(size_t i = 0; i < viewportNum; i++) {
+        mtlViewports[i].originX = viewports[i].x;
+        mtlViewports[i].originY = viewports[i].y;
+        mtlViewports[i].width = viewports[i].width;
+        mtlViewports[i].height = viewports[i].height;
+        mtlViewports[i].znear = viewports[i].depthMin;
+        mtlViewports[i].zfar = viewports[i].depthMax;
+    }
+    [m_RendererEncoder setViewports: mtlViewports count: viewportNum];
+}
+
+
+void CommandBufferMTL::SetScissors(const Rect* rects, uint32_t rectNum) {
+    NSCAssert(m_RendererEncoder, @"encoder set");
+    MTLScissorRect rect;
+    rect.x = rects[rectNum].x;
+    rect.y = rects[rectNum].y;
+    rect.width = rects[rectNum].width;
+    rect.height = rects[rectNum].height;
+    [m_RendererEncoder setScissorRect: rect];
+}
+void CommandBufferMTL::SetDepthBounds(float boundsMin, float boundsMax) {
+    //[m_RendererEncoder set]
+}
+void CommandBufferMTL::SetStencilReference(uint8_t frontRef, uint8_t backRef) {
+    [m_RendererEncoder setStencilFrontReferenceValue: frontRef backReferenceValue:backRef];
+}
+
+void CommandBufferMTL::SetBlendConstants(const Color32f& color) {
+    [m_RendererEncoder
+     setBlendColorRed:color.x
+     green:color.y 
+     blue:color.z
+     alpha:color.w
+    ];
+}
+void CommandBufferMTL::SetShadingRate(const ShadingRateDesc& shadingRateDesc) {
+
+    //[m_RendererEncoder sha]
+}
+
+void CommandBufferMTL::ClearAttachments(const ClearDesc* clearDescs, uint32_t clearDescNum, const Rect* rects, uint32_t rectNum) {
+  
+
+    Restore();
+}
```

Comment:
there isn't a similar implementation in metal for this. umm I guess it would have to cache out a pipeline and render rectangles to the various layers then restore some of the command buffer state?

Reply:
Probably you could look at MoltenVK (VK to Metal) for `vkCmdClearAttachments`.

P.S. (since I'm here) Really appreciate your work!

Reply:
https://github.com/KhronosGroup/MoltenVK/blob/main/MoltenVK/MoltenVK/Commands/MVKCmdTransfer.mm#L1275

Reply:
I did see that umm, that does make this complicated. i guess i have to cache all the pipeline state and submit it in one batch. I think you can only have one encoder at any one time for a cmd buffer. does make the data structure a bit fat.

Reply:
You can try suggesting API changes simplifying Metal implementation. Very undesired to be honest :)

Reply:
at the moment im tracking any state for the vieport and only start a rendering encoder when a pipeline is set. you can have a look at the changes.